### DISCUSSION TITLE
torch/nn/init.py ZeroDivisionError

### DISCUSSION THREAD

**Discussion by luc-henry** [neuralhydrology/neuralhydrology#32] [Q&A] [answered]
Hi @kratzert ,
Sorry to disturb you again.  Do you have any idea why the torch init.py. could return a zero division error for 
`line 376, in kaiming_uniform_
    std = gain / math.sqrt(fan)`

**Comment by kratzert** [comment]
Hi @henry-wni 

from the little background you provide, it is hard to tell what the exact reason of that error is or where it is occurring. For the future, it is best to copy the entire error message stack (so that we can see which file is causing the error) + eventually the content of your config file.

However, I think the problem is similar to last time in that you are trying to use something without inputs. E.g. the EA-LSTM model can only be used for multi-basin training with static features or basin one-hot encoding (we should probably raise in error for that case). Could that be the issue?

**Reply by luc-henry** [reply]
Thanks a lot that was it. 
Got an issue with the 
`  File "/home/ec2-user/SageMaker/neuralhydrology/neuralhydrology/datasetzoo/genericdataset.py", line 63, in __init__
    scaler=scaler)
  File "/home/ec2-user/SageMaker/neuralhydrology/neuralhydrology/datasetzoo/basedataset.py", line 128, in __init__
    self._load_data()
  File "/home/ec2-user/SageMaker/neuralhydrology/neuralhydrology/datasetzoo/basedataset.py", line 517, in _load_data
    xr = self._load_or_create_xarray_dataset()
  File "/home/ec2-user/SageMaker/neuralhydrology/neuralhydrology/datasetzoo/basedataset.py", line 282, in _load_or_create_xarray_dataset
    native_frequency = utils.infer_frequency(df.index)
  File "/home/ec2-user/SageMaker/neuralhydrology/neuralhydrology/datautils/utils.py", line 170, in infer_frequency
    native_frequency = pd.infer_freq(index)
  File "/home/ec2-user/anaconda3/envs/JupyterSystemEnv/lib/python3.6/site-packages/pandas/tseries/frequencies.py", line 175, in infer_freq
    f"cannot infer freq from a non-convertible index type {type(index)}"
TypeError: cannot infer freq from a non-convertible index type <class 'pandas.core.indexes.numeric.Int64Index'>` now

**Reply by luc-henry** [reply]
Hi, @kratzert  @gauchm 
Sorry to disturb again.  Am I doing something wrong with the Datetime ? I still have this issue with the` fix-frequencies` and generic dataset

**Reply by kratzert** [reply]
I think the issue here is that your dataframe is not datetime indexed but that your index is an integer (see last message of the errror text).  Make sure that if you load on the of the netCDF files manually by:

```python
xr = xarray.open_dataset("/path/to/netcdf_file.nc4")
df = xr.to_dataframe()
```
That the index of the dataframe is a datetimeindex instance.

**Reply by luc-henry** [reply]
I need both a datetimeindex and a ['date'] datetime  ?
I check and the type seems correct. 
`  File "/home/ec2-user/SageMaker/neuralhydrology/neuralhydrology/datasetzoo/basedataset.py", line 281, in _load_or_create_xarray_dataset
    native_frequency = utils.infer_frequency(df.index)
  File "/home/ec2-user/SageMaker/neuralhydrology/neuralhydrology/datautils/utils.py", line 165, in infer_frequency
    raise ValueError(f'Cannot infer a legal frequency from dataset: {native_frequency}.')
ValueError: Cannot infer a legal frequency from dataset: None.`

**Comment by kratzert** [comment]
What you need is that the index of your dataframe that results of the two lines I posted above is a pandas datetimeindex. The name of the index should be "datetime" if I am not wrong. It is still the same issue that something with your index is not correct. Can you dump a single netcdf file somewhere?

**Reply by luc-henry** [reply]
Thank you. I have a  DatetimeIndex (dtype='datetime64[ns]', name='date', length=309807, freq=None). But still have issue.

**Reply by kratzert** [reply]
You can see your problem right there: `freq=None`. Is it possible that your dataframe is irregularly spaced? Then pandas can't infer frequency. You could do a `df.reindex` using `pd.date_range(df.index[0], df.index[-1], freq=freq)`, where `freq` is the frequency string that you think your data should have, e.g. `1H` for hourly or `1D` for daily. Missing dates will be filled with NaNs, which is okay.

**Reply by luc-henry** [reply]
Can I have several  same range of dates for several coordinates in a same netcdf file for one basin ?

**Reply by kratzert** [reply]
Not sure if I understand what you mean. Can you try to explain it with an example and/or upload a sample netcdf? Usually your netcdf file for one basin should have one coordinate (the datetimeindex) as a sequential index without gaps and the a varying number of data variables.

**Reply by luc-henry** [reply]
Sorry. Thanks a lot. It looks like this 
[data_exemple.pdf](https://github.com/neuralhydrology/neuralhydrology/files/6243029/data_exemple.pdf)

**Reply by kratzert** [reply]
Uff any chance you can send me a netCDF and not a pdf? I can't look into the data, having a pdf table. Why is the first row having an hour `00:00` and the others only a date?

**Reply by luc-henry** [reply]
I am very sorry for the annoying situation and thank you very much for the help.  I will try to send you what I can tomorrow.

**Comment by kratzert** [comment, marked as answer]
Thanks for sending me the netCDF file. I think there are multiple issues with your data and the generic dataset class you are trying to use:

- You xarray is multi-indexed (lat, lon, time), which is not supported by the GenericDataset. This dataset expects one netcdf file per "entity" (e.g. basin) that has a single index, which is the datetime. You can work with your netcdf file,  which will require you to modify the genericdataset and essentially write your own custom dataset (see the tutorial in the documentation). Remember, the return value of the dataset methods should be a time-indexed dataframe for one "entity" (basin or grid cell).
- Your datetime index is not continuous. The length of the xarray is 34423, while the regular date range between the first and last date is 35857 (`len(pd.date_range(df.index[0], df.index[-1], freq="1H"))`). This is not supported by `pd.infer_freq` which we use in the code to get the frequency of the input data.

So there are two things you have to do:
1. Make sure your xarray file has a continuous time index and fill missing dates e.g. with NaNs
2. Either store the time series for each gridcell in a separate netcdf file that is only indexed by datetime (then you can use the GenericDataset) or write your own dataset class (see tutorial) that is able to work with the multi-indexed xarray and only returns the timeseries for one grid cell at the time.

**Reply by luc-henry** [reply]
Thank you for taking the time and for the explanations
